middleware/opencensus_tracing.go

```diff
@@ -2,50 +2,99 @@ package middleware
 
 import (
 	"context"
+	"crypto/rand"
 	"encoding/base64"
 	"fmt"
+	"math/big"
 	"net/http"
+	"strconv"
 
 	"go.opencensus.io/trace"
 	"go.opencensus.io/trace/propagation"
 )
 
-const HeaderNameOpencensusSpan = "X-Opencensus-Span"
+const (
+	headerNameOpencensusSpan           = "X-Opencensus-Span"
+	headerNameOpencensusSpanEventIDKey = "X-Opencensus-event-id"
+	spanRequestPayloadAttributeKey     = "event-payload"
+	spanResponsePayloadAttributeKey    = "event-response"
+)
 
 // AddTracingSpanToRequest resolves span data from the provided context and injects it to the request
-func AddTracingSpanToRequest(ctx context.Context, req *http.Request) {
+func AddTracingSpanToRequest(ctx context.Context, r *http.Request) {
+	body := decorateRequestBody(r)
+	r.Body = body
+
 	span := trace.FromContext(ctx)
 	if span == nil {
 		return
 	}
-	setSpanHeader(span.SpanContext(), req)
+
+	eID := generateEventID()
+	eIDString := strconv.FormatInt(eID, 10)
+	r.Header.Set(headerNameOpencensusSpanEventIDKey, eIDString)
+
+	// TODO: test it!
+	span.AddMessageSendEvent(eID, r.ContentLength, 0)
+	span.AddAttributes(trace.StringAttribute(spanRequestPayloadAttributeKey, string(body.Payload())))
+
+	setSpanHeader(span.SpanContext(), r)
 }
 
 // OpencensusTracing implements a simple middleware handler
 // for adding an opencensus tracing span to the request context
 func OpencensusTracing() func(next http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
 		fn := func(w http.ResponseWriter, r *http.Request) {
+			ww := decorateResponseWriter(w)
+
+			body := decorateRequestBody(r)
+			r.Body = body
+
 			ctx := r.Context()
+			var span *trace.Span
 
 			parentSpanContext, ok := getSpanContext(r)
 			if ok {
-				c, span := trace.StartSpanWithRemoteParent(ctx, spanName(r), parentSpanContext)
+				ctx, span = trace.StartSpanWithRemoteParent(ctx, spanName(r), parentSpanContext)
 				span.AddLink(trace.Link{
 					TraceID:    parentSpanContext.TraceID,
 					SpanID:     parentSpanContext.SpanID,
 					Type:       trace.LinkTypeParent,
 					Attributes: nil,
 				})
-				span.End()
-				ctx = c
 			} else {
-				c, span := trace.StartSpan(ctx, spanName(r))
-				span.End()
-				ctx = c
+				ctx, span = trace.StartSpan(ctx, spanName(r))
 			}
 
-			next.ServeHTTP(w, r.WithContext(ctx))
+			defer func() {
```

Comment:
I guess that it may be nice to move it to separate function (as it's pretty deep with indents)

Reply:
Done